src/Microsoft.ApiDesignGuidelines.Analyzers/Core/PassSystemUriObjectsInsteadOfStrings.cs

```diff
@@ -24,15 +29,191 @@ public abstract class PassSystemUriObjectsInsteadOfStringsAnalyzer : DiagnosticA
-                                                                             isEnabledByDefault: false,
+                                                                             isEnabledByDefault: true,
                                                                              description: s_localizableDescription,
-                                                                             helpLinkUri: null,     // TODO: add MSDN url
+                                                                             helpLinkUri: "https://msdn.microsoft.com/en-us/library/ms182360.aspx",
                                                                              customTags: WellKnownDiagnosticTags.Telemetry);
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
 
         public override void Initialize(AnalysisContext analysisContext)
         {
+            // this is stateless analyzer, can run concurrently
+            analysisContext.EnableConcurrentExecution();
+
+            // this has no meaning on running on generated code which user can't control
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterCompilationStartAction(c =>
+            {
+                var analyzer = new Analyzer(c.Compilation, GetInvocationExpression);
+                c.RegisterOperationAction(analyzer.Analyze, OperationKind.InvocationExpression);
+            });
+        }
+
+        protected abstract SyntaxNode GetInvocationExpression(SyntaxNode invocationNode);
+
+        private struct Analyzer
+        {
+            private static readonly string[] uriWords = new string[] { "uri", "urn", "url" };
+
+#pragma warning disable RS1008 // this type will be created per compilation
+            private readonly Compilation _compilation;
+            private readonly INamedTypeSymbol _string;
+            private readonly INamedTypeSymbol _uri;
+            private readonly Func<SyntaxNode, SyntaxNode> _expressionGetter;
+#pragma warning restore RS1008
+
+            public Analyzer(Compilation compilation, Func<SyntaxNode, SyntaxNode> expressionGetter)
+            {
+                _compilation = compilation;
+                _string = WellKnownTypes.String(compilation);
+                _uri = WellKnownTypes.Uri(compilation);
+
+                _expressionGetter = expressionGetter;
+            }
+
+            public void Analyze(OperationAnalysisContext context)
+            {
+                if (context.Operation.IsInvalid)
+                {
+                    // not interested in invalid expression
+                    return;
+                }
+
+                var invocation = (IInvocationExpression)context.Operation;
+                var method = invocation.TargetMethod;
+
+                var node = context.Operation.Syntax;
+                if (node == null || node.SyntaxTree == null)
```

Comment:
We would want the operation (that based on compiler generated bound node) to have a non-null Syntax property as well, typically by providing a parent syntax node of some sort. So if you encountered any operation with null Syntax, it would be a bug. Could you please file an issue for that?